pdns/recursordist/recursor_cache.cc

```diff
@@ -211,11 +214,22 @@ time_t MemRecursorCache::handleHit(time_t now, MapCombo::LockedContent& content,
   }
 
   if (signatures != nullptr) {
-    signatures->insert(signatures->end(), entry->d_signatures.begin(), entry->d_signatures.end());
+    if (*signatures && !(*signatures)->empty() && entry->d_signatures && !entry->d_signatures->empty()) {
+      // Return a new vec if we need to append to a non-empty vector
+      SigRecsVec vec(**signatures);
+      vec.insert(vec.end(), entry->d_signatures->cbegin(), entry->d_signatures->cend());
+      *signatures = std::make_shared<SigRecsVec>(vec);
+    }
+    else {
+      assert(*signatures == nullptr || (*signatures)->empty());
```

Comment:
This feels wrong. If the condition above is not correct, we could get here with some entries in `*signatures` as long as `entry->d_signatures` doesn't contain any, right?